pkg/apis/traceflow/v1/types.go

```diff
@@ -0,0 +1,162 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type DropReason string
+
+type Phase string
+
+const (
+	INITIAL Phase = "INITIAL"
+	RUNNING Phase = "RUNNING"
+	SUCCESS Phase = "SUCCESS"
+	TIMEOUT Phase = "TIMEOUT"
+	ERROR   Phase = "ERROR"
+)
+
+type Component string
+
+const (
+	SPOOFGUARD    Component = "SPOOFGUARD"
+	LB            Component = "LB"
+	ROUTING       Component = "ROUTING"
+	NETWORKPOLICY Component = "NETWORKPOLICY"
+	FORWARDING    Component = "FORWARDING"
+)
+
+type Action string
+
+const (
+	DELIVERED Action = "DELIVERED"
+	RECEIVED  Action = "RECEIVED"
+	FORWARDED Action = "FORWARDED"
+	DROPPED   Action = "DROPPED"
+)
+
+type PacketResouceType string
+
+const (
+	FieldsPacketData PacketResouceType = "FieldsPacketData"
+)
+
+type PacketTransportType string
+
+const (
+	UNICAST PacketTransportType = "UNICAST"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type Traceflow struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Source      `json:"source"`
+	Destination `json:"destination"`
+	RoundID     string `json:"roundID,omitempty"`
+
+	Packet `json:"packet"`
+	Status `json:"status"`
+}
+
+type Source struct {
+	SrcNamespace string `json:"srcNamespace,omitempty"`
+	SrcPod       string `json:"srcPod,omitempty"`
+}
+
+type Destination struct {
+	DstNamespace string `json:"dstNamespace,omitempty"`
+	DstPod       string `json:"dstPod,omitempty"`
+	DstService   string `json:"dstService,omitempty"`
+	DstIP        string `json:"dstIP,omitempty"`
+}
+
+type IPHeader struct {
+	SrcIP    string `json:"srcIP,omitempty"`
+	Protocol int32  `json:"protocol,omitempty"`
+	TTL      int32  `json:"ttl,omitempty"`
+	Flags    int32  `json:"flags,omitempty"`
+}
+
+type TransportHeader struct {
+	ICMPEchoRequestHeader `json:"icmpEchoRequestHeader"`
+	UDPHeader             `json:"udpHeader"`
+	TCPHeader             `json:"tcpHeader"`
+}
+
+type ICMPEchoRequestHeader struct {
+	ID       int32 `json:"id,omitempty"`
+	Sequence int32 `json:"sequence,omitempty"`
+}
+
+type UDPHeader struct {
+	SrcUDPPort int32 `json:"srcUDPPort,omitempty"`
+	DstUDPPort int32 `json:"dstUDPPort,omitempty"`
+}
+
+type TCPHeader struct {
+	SrcTCPPort int32 `json:"srcTCPPort,omitempty"`
+	DstTCPPort int32 `json:"dstTCPPort,omitempty"`
+	TCPFlags   int32 `json:"tcpFlags,omitempty"`
+}
+
+type Packet struct {
+	ResouceType   PacketResouceType   `json:"resouceType,omitempty"`
+	TransportType PacketTransportType `json:"transportType,omitempty"`
+	PayloadString string              `json:"payloadString,omitempty"`
+	PayloadSize   int32               `json:"payloadSize,omitempty"`
+
+	IPHeader        `json:"ipHeader"`
+	TransportHeader `json:"transportHeader"`
+}
+
+type Status struct {
+	Phase        Phase         `json:"phase,omitempty"`
+	CrossNodeTag uint8         `json:"crossNodeTag,omitempty"`
+	NodeSender   []Observation `json:"nodeSender,omitempty"`
```

Comment:
Is that we assume a single sender and a single receiver? I think later we might have cases a packet goes through more than two nodes. Since Observation already includes Node information, should we just use a single Observation slice for all Nodes?
If we want to group Observations by Nodes, I would suggest to add another struct like NodeObservations that includes the Node information, and a slice of Observations.

Reply:
I prefer not to use single observation slice for all nodes, because the sequence may not be expected. Meanwhile, we need to confirm if it is safe to add objects to one slice from multiple clients in parallel.
Because we don't have multicast now, we can assume a single sender and a single receiver. If packet will go to third node, we can add another node with different role.

Reply:
Then could we either use a single slice for all Nodes, or use a slice of a new Node struct, which includes the Node info (Node name) and a slice of Observation (with NodeName moves to the Node struct).
The latter is like:
`
    NodeObservations []NodeObservation

type NodeObservation struct {
     NodeName string
     Observations []Observation    
}
`

Reply:
Latter one sounds applicable, but still need to verify if it is safe to add objects to one slice from multiple clients in parallel.
This operation is only supported by JSON patch, which is not tested in Antrea before. Currently we use Strategic merge patch to nodeSender/nodeReceiver by different nodes.

Reply:
Ok. Will wait for your verification.

Reply:
Have we figured out if we can use a single Node slice or not? I do not like the current definition, because 1) later we can have more than two hops; 2) a packet can have multiple receivers (e.g. multicast).
Even we have to use two field, I would call it "SenderNode" or "OriginNode" / "ReceiverNode" or "DestinationNode".